{
  "metadata": {
    "kernelspec": {
      "name": "python",
      "display_name": "Python (Pyodide)",
      "language": "python"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "python",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8"
    }
  },
  "nbformat_minor": 5,
  "nbformat": 4,
  "cells": [
    {
      "id": "03991e52-477d-4098-960c-81d4cd67a702",
      "cell_type": "code",
      "source": "## Early access chromatin stenciling-tuned basecalling models\n\nThese basecalling models have been tested and shown to have better performance on chromatin stenciling (exogenous 6mA-treated) DNA reads than the release models. Background 6mA levels do have an inverse relationship with accuracy, however.\\\nThese models should work with dorado v0.8+.\n\nRecommended modified base models:\n\n#6mA_sup:\nres_dna_r10.4.1_e8.2_400bps_sup@20241218_chromatin_stenciling: \ndna_r10.4.1_e8.2_400bps_sup@v5.0.0_6mA@v3 \nor dna_r10.4.1_e8.2_400bps_sup@v5.0.0_6mA@v1\n\n#6mA_hac:\nres_dna_r10.4.1_e8.2_400bps_hac@20241218_chromatin_stenciling: \ndna_r10.4.1_e8.2_400bps_hac@v5.0.0_6mA@v3\n\n#5mC/5hmC sup:\nres_dna_r10.4.1_e8.2_400bps_sup@20241218_chromatin_stenciling: \ndna_r10.4.1_e8.2_400bps_sup@v5.0.0_5mCG_5hmCG@v3 \nor dna_r10.4.1_e8.2_400bps_sup@v5.0.0_5mCG_5hmCG@v2.0.1\\\n\n#5mC/5hmC hac:\nres_dna_r10.4.1_e8.2_400bps_hac@20241218_chromatin_stenciling: \ndna_r10.4.1_e8.2_400bps_hac@v5.0.0_5mCG_5hmCG@v3 \nor dna_r10.4.1_e8.2_400bps_hac@v5.0.0_5mCG_5hmCG@v2\\\n}",
      "metadata": {
        "trusted": true
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "id": "6961b47c-3562-4bb2-b18c-8a04c56fc6c8",
      "cell_type": "code",
      "source": "# Calculate enrichment of 6mA in chromatin stenciling experiments\n\nThis small project contains scripts and resources for calculating the peak over background levels of 6mA in chromatin stenciling (fiber-seq) experiments.\nTo perform the evaluation, there are three steps:\n\n## Quick start\n\n0. Install a python environment based on the requirements in `requirements.txt`, there is a helper script (`install_venv.sh`) that you may use.\n1. Run `$ bash run_6mA_peak_over_background.sh` running this script with no arguments will display a help and description of the input files required.\n2. Run `$ python calculate_snr.py ${localize_tsv} [--outdir ${output_directory}]` to calculate the metric. Optionally output a new plot.\n\n## Description of metric\n\nThe output metric peak_over_background the ratio of the 6mA levels centered at housekeeping gene promoters. \nThe housekeeping genes are culled from the [HRT atlas](https://housekeeping.unicamp.br/) then joined with promoters from the [Eukaryotic promoter database](https://epd.expasy.org/epd).\n\nBelow is an example plot generated by `modkit localize`:\n\n![signal-to-noise](signal-to-noise-desc.png)\n\nThe metric is then a ratio of the value of the blue line and the red line.\n\n## Notes\n\n- For Human runs, align your reads to hg38\n- For Mouse runs, align your reads to mm10\n- Use `dorado aligner` to map reads (that aren't mapped during basecalling) to make sure that modified base tags aren't corrupted or lost\n",
      "metadata": {
        "trusted": true
      },
      "outputs": [],
      "execution_count": null
    }
  ]
}