{
  "metadata": {
    "kernelspec": {
      "name": "python",
      "display_name": "Python (Pyodide)",
      "language": "python"
    },
    "language_info": {
      "name": ""
    }
  },
  "nbformat_minor": 5,
  "nbformat": 4,
  "cells": [
    {
      "id": "9d24c2a6-76cc-4794-8afd-c06fdec09d87",
      "cell_type": "markdown",
      "source": "# modkit extract full -h\n\nTransform the probabilities from the MM/ML tags in a modBAM into a table\nUsage: modkit extract full [OPTIONS] <IN_BAM> <OUT_PATH>\n\nArguments:\n  <IN_BAM>    Path to modBAM file to extract read-level information from, or one\n              of `-` or `stdin` to specify a stream from standard input. If a\n              file is used it may be sorted and have associated index\n  <OUT_PATH>  Path to output file, \"stdout\" or \"-\" will direct output to\n              standard out\n\nOptions:\n      --reference <REFERENCE>  Path to reference FASTA to extract reference\n                               context information from. Required for motif\n                               selection\n  -h, --help                   Print help (see more with '--help')\n\nCompute Options:\n  -t, --threads <THREADS>\n          Number of threads to use [default: 4]\n      --out-threads <OUT_THREADS>\n          Number of threads to use for parallel bgzf writing [default: 4]\n  -q, --queue-size <QUEUE_SIZE>\n          Number of reads that can be in memory at a time. Increasing this value\n          will increase thread usage, at the cost of memory usage [default:\n          10000]\n\nOutput Options:\n      --bgzf                   Write output as BGZF compressed file\n      --force                  Force overwrite of output file\n      --kmer-size <KMER_SIZE>  Set the query and reference k-mer size (if a\n                               reference is provided). Maximum number for this\n                               value is 50 [default: 5]\n      --no-headers             Don't print the header lines in the output tables\n\nLogging Options:\n      --log-filepath <LOG_FILEPATH>  Path to file to write run log\n\nSelection Options:\n      --mapped-only\n          Include only mapped bases in output (alias: mapped)\n      --allow-non-primary\n          Output aligned secondary and supplementary base modification\n          probabilities as additional rows. The primary alignment will have all\n          of the base modification probabilities (including soft-clipped ones,\n          unless --mapped-only is used). The non-primary alignments will only\n          have mapped bases in the output\n      --num-reads <NUM_READS>\n          Number of reads to use. Note that when using a sorted, indexed modBAM\n          that the sampling algorithm will attempt to sample records evenly over\n          the length of the reference sequence. The result is the final number\n          of records used may be slightly more or less than the requested\n          number. When piping from stdin or using a modBAM without an index, the\n          requested number of reads will be the first `num_reads` records\n      --region <REGION>\n          Process only reads that are aligned to a specified region of the BAM.\n          Format should be <chrom_name>:<start>-<end> or <chrom_name>\n      --include-bed <INCLUDE_BED>\n          BED file with regions to include (alias: include-positions).\n          Implicitly only includes mapped sites\n  -v, --exclude-bed <EXCLUDE_BED>\n          BED file with regions to _exclude_ (alias: exclude)\n      --edge-filter <EDGE_FILTER>\n          Discard base modification calls that are this many bases from the\n          start or the end of the read. Two comma-separated values may be\n          provided to asymmetrically filter out base modification calls from the\n          start and end of the reads. For example, 4,8 will filter out base\n          modification calls in the first 4 and last 8 bases of the read\n      --invert-edge-filter\n          Invert the edge filter, instead of filtering out base modification\n          calls at the ends of reads, only _keep_ base modification calls at the\n          ends of reads. E.g. if usually, \"4,8\" would remove (i.e. filter out)\n          base modification calls in the first 4 and last 8 bases of the read,\n          using this flag will keep only base modification calls in the first 4\n          and last 8 bases\n\nModified Base Options:\n      --motif <MOTIF> <MOTIF>  Output read-level base modification probabilities\n                               restricted to the reference sequence motifs\n                               provided. The first argument should be the\n                               sequence motif and the second argument is the\n                               0-based offset to the base to pileup base\n                               modification counts for. For example: --motif\n                               CGCG 0 indicates include base modifications for\n                               which the read is aligned to the first C on the\n                               top strand and the last C (complement to G) on\n                               the bottom strand. The --cpg argument is short\n                               hand for --motif CG 0. This argument can be\n                               passed multiple times\n      --annotate-motifs        When used with `--motif` or `--cpg` emit all\n                               modified base alignment information even if it\n                               does not align to a reference motif, but annotate\n                               which aligned positions match which motifs in the\n                               \"motifs\" column. \".\" will be used when an aligned\n                               position does not match a motif\n      --cpg                    Only output counts at CpG motifs. Requires a\n                               reference sequence to be provided\n",
      "metadata": {}
    },
    {
      "id": "61700d52-e24a-4445-b5fa-4a1b5cfa8149",
      "cell_type": "code",
      "source": "",
      "metadata": {
        "trusted": true
      },
      "outputs": [],
      "execution_count": null
    },
    {
      "id": "3b3b03ad-2fdc-45cb-9c75-6e87232c7cbf",
      "cell_type": "markdown",
      "source": "## We do modkit extract calls with filter-threshold and cpg just to get CpG methylated C with a threshold of more than 0.80. Then we generate a python script that cound the number of times that each read appear on this file and generate another .tsv file with the amount of m for each read on the region of the BED file. Pay attention to the fact that cutted reads will have fewer m, but for this is important to keep the right cut site on the reference.\n\nmodkit extract calls \\\n./data/IE20_multiplex/IE20_barcode03/IE20_barcode03_lowfiltered_alignedcutq4.bam \\\n./output/IE20_multiplex/IE20_barcode03/modkit_extractcalls/IE20_barcode03_q4/IE20_barcode03_extractcalls_3000q4.tsv \\\n--reference ./data/reference_genomes/cut_hg002v1.1.fasta \\\n--include-bed ./data/reference_regions/CpG_island_subtelomeres_3000.bed \\\n--filter-threshold C:0.80 \\\n--cpg \n\n\n# modkit extract calls -h\nProduce a table of read-level base modification calls. This table has, for each\nread, one row for each base modification call in that read using the same\nthresholding algorithm as in pileup, or summary (see online documentation for\ndetails on thresholds)\n\nUsage: modkit extract calls [OPTIONS] <IN_BAM> <OUT_PATH>\n\nArguments:\n  <IN_BAM>    Path to modBAM file to extract read-level information from, or one\n              of `-` or `stdin` to specify a stream from standard input. If a\n              file is used it may be sorted and have associated index\n  <OUT_PATH>  Path to output file, \"stdout\" or \"-\" will direct output to\n              standard out\n\nOptions:\n      --reference <REFERENCE>  Path to reference FASTA to extract reference\n                               context information from. If no reference is\n                               provided, `ref_kmer` column will be \".\" in the\n                               output. (alias: ref)\n  -h, --help                   Print help (see more with '--help')\n\nCompute Options:\n  -t, --threads <THREADS>\n          Number of threads to use [default: 4]\n      --out-threads <OUT_THREADS>\n          Number of threads to use for parallel bgzf writing [default: 4]\n  -q, --queue-size <QUEUE_SIZE>\n          Number of reads that can be in memory at a time. Increasing this value\n          will increase thread usage, at the cost of memory usage [default:\n          10000]\n\nOutput Options:\n      --bgzf                   Write output as BGZF compressed file\n      --force                  Force overwrite of output file\n      --kmer-size <KMER_SIZE>  Set the query and reference k-mer size (if a\n                               reference is provided). Maximum number for this\n                               value is 50 [default: 5]\n      --no-headers             Don't print the header lines in the output tables\n\nLogging Options:\n      --log-filepath <LOG_FILEPATH>  Path to file to write run log\n\nSelection Options:\n      --mapped-only\n          Include only mapped bases in output (alias: mapped)\n      --allow-non-primary\n          Output aligned secondary and supplementary base modification\n          probabilities as additional rows. The primary alignment will have all\n          of the base modification probabilities (including soft-clipped ones,\n          unless --mapped-only is used). The non-primary alignments will only\n          have mapped bases in the output\n      --num-reads <NUM_READS>\n          Number of reads to use. Note that when using a sorted, indexed modBAM\n          that the sampling algorithm will attempt to sample records evenly over\n          the length of the reference sequence. The result is the final number\n          of records used may be slightly more or less than the requested\n          number. When piping from stdin or using a modBAM without an index, the\n          requested number of reads will be the first `num_reads` records\n      --region <REGION>\n          Process only reads that are aligned to a specified region of the BAM.\n          Format should be <chrom_name>:<start>-<end> or <chrom_name>\n      --include-bed <INCLUDE_BED>\n          BED file with regions to include (alias: include-positions).\n          Implicitly only includes mapped sites\n  -v, --exclude-bed <EXCLUDE_BED>\n          BED file with regions to _exclude_ (alias: exclude)\n      --edge-filter <EDGE_FILTER>\n          Discard base modification calls that are this many bases from the\n          start or the end of the read. Two comma-separated values may be\n          provided to asymmetrically filter out base modification calls from the\n          start and end of the reads. For example, 4,8 will filter out base\n          modification calls in the first 4 and last 8 bases of the read\n      --invert-edge-filter\n          Invert the edge filter, instead of filtering out base modification\n          calls at the ends of reads, only _keep_ base modification calls at the\n          ends of reads. E.g. if usually, \"4,8\" would remove (i.e. filter out)\n          base modification calls in the first 4 and last 8 bases of the read,\n          using this flag will keep only base modification calls in the first 4\n          and last 8 bases\n      --pass-only\n          Only output base modification calls that pass the minimum confidence\n          threshold. (alias: pass)\n\nModified Base Options:\n      --motif <MOTIF> <MOTIF>  Output read-level base modification probabilities\n                               restricted to the reference sequence motifs\n                               provided. The first argument should be the\n                               sequence motif and the second argument is the\n                               0-based offset to the base to pileup base\n                               modification counts for. For example: --motif\n                               CGCG 0 indicates include base modifications for\n                               which the read is aligned to the first C on the\n                               top strand and the last C (complement to G) on\n                               the bottom strand. The --cpg argument is short\n                               hand for --motif CG 0. This argument can be\n                               passed multiple times\n      --annotate-motifs        When used with `--motif` or `--cpg` emit all\n                               modified base alignment information even if it\n                               does not align to a reference motif, but annotate\n                               which aligned positions match which motifs in the\n                               \"motifs\" column. \".\" will be used when an aligned\n                               position does not match a motif\n      --cpg                    Only output counts at CpG motifs. Requires a\n                               reference sequence to be provided\n\nFiltering Options:\n      --filter-threshold <FILTER_THRESHOLD>\n          Specify the filter threshold globally or per-base. Global filter\n          threshold can be specified with by a decimal number (e.g. 0.75).\n          Per-base thresholds can be specified by colon-separated values, for\n          example C:0.75 specifies a threshold value of 0.75 for cytosine\n          modification calls. Additional per-base thresholds can be specified by\n          repeating the option: for example --filter-threshold C:0.75\n          --filter-threshold A:0.70 or specify a single base option and a\n          default for all other bases with: --filter-threshold A:0.70\n          --filter-threshold 0.9 will specify a threshold value of 0.70 for\n          adenine and 0.9 for all other base modification calls\n\nSampling Options:\n  -n, --sample-num-reads <SAMPLE_NUM_READS>\n          Sample this many reads when estimating the filtering threshold. If a\n          sorted, indexed modBAM is provided reads will be sampled evenly across\n          aligned genome. If a region is specified, with the --region, then\n          reads will be sampled evenly across the region given. This option is\n          useful for large BAM files. In practice, 10-50 thousand reads is\n          sufficient to estimate the model output distribution and determine the\n          filtering threshold [default: 10042]\nTransform the probabilities from the MM/ML tags in a modBAM into a table\n",
      "metadata": {}
    },
    {
      "id": "0b345dbc-7038-4b73-b7d4-7bf904e31184",
      "cell_type": "markdown",
      "source": "# modkit pileup -h\n\nmodkit pileup path/to/reads.bam output/path/pileup.bed --ref path/to/reference.fasta\n\nTabulates base modification calls across genomic positions. This command\nproduces a bedMethyl formatted file. Schema and description of fields can be\nfound in the README\n\nUsage: modkit pileup [OPTIONS] <IN_BAM> <OUT_BED>\n\nArguments:\n  <IN_BAM>   Input BAM, should be sorted and have associated index available\n  <OUT_BED>  Output file (or directory with --bedgraph option) to write results\n             into. Specify \"-\" or \"stdout\" to direct output to stdout\n\nOptions:\n      --preset <PRESET>     Optional preset options for specific applications.\n                            traditional: Prepares bedMethyl analogous to that\n                            generated from other technologies for the analysis\n                            of 5mC modified bases. Shorthand for --cpg\n                            --combine-strands --ignore h [possible values:\n                            traditional]\n      --invert-edge-filter  Invert the edge filter, instead of filtering out\n                            base modification calls at the ends of reads, only\n                            _keep_ base modification calls at the ends of reads.\n                            E.g. if usually, \"4,8\" would remove (i.e. filter\n                            out) base modification calls in the first 4 and last\n                            8 bases of the read, using this flag will keep only\n                            base modification calls in the first 4 and last 8\n                            bases\n  -h, --help                Print help (see more with '--help')\n\nLogging Options:\n      --log-filepath <LOG_FILEPATH>\n          Specify a file for debug logs to be written to, otherwise ignore them.\n          Setting a file is recommended. (alias: log)\n\nSelection Options:\n      --region <REGION>  Process only the specified region of the BAM when\n                         performing pileup. Format should be\n                         <chrom_name>:<start>-<end> or <chrom_name>. Commas are\n                         allowed\n\nCompute Options:\n  -t, --threads <THREADS>  Number of threads to use while processing chunks\n                           concurrently [default: 4]\n\nSampling Options:\n  -n, --num-reads <NUM_READS>  Sample this many reads when estimating the\n                               filtering threshold. Reads will be sampled evenly\n                               across aligned genome. If a region is specified,\n                               either with the --region option or the\n                               --sample-region option, then reads will be\n                               sampled evenly across the region given. This\n                               option is useful for large BAM files. In\n                               practice, 10-50 thousand reads is sufficient to\n                               estimate the model output distribution and\n                               determine the filtering threshold [default:\n                               10042]\n\nFiltering Options:\n      --no-filtering\n          Do not perform any filtering, include all mod base calls in output.\n          See filtering.md for details on filtering\n      --filter-threshold <FILTER_THRESHOLD>\n          Specify the filter threshold globally or per-base. Global filter\n          threshold can be specified with by a decimal number (e.g. 0.75).\n          Per-base thresholds can be specified by colon-separated values, for\n          example C:0.75 specifies a threshold value of 0.75 for cytosine\n          modification calls. Additional per-base thresholds can be specified by\n          repeating the option: for example --filter-threshold C:0.75\n          --filter-threshold A:0.70 or specify a single base option and a\n          default for all other bases with: --filter-threshold A:0.70\n          --filter-threshold 0.9 will specify a threshold value of 0.70 for\n          adenine and 0.9 for all other base modification calls\n      --mod-thresholds <MOD_THRESHOLDS>\n          Specify a passing threshold to use for a base modification,\n          independent of the threshold for the primary sequence base or the\n          default. For example, to set the pass threshold for 5hmC to 0.8 use\n          `--mod-threshold h:0.8`. The pass threshold will still be estimated as\n          usual and used for canonical cytosine and other modifications unless\n          the `--filter-threshold` option is also passed. See the online\n          documentation for more details\n      --sample-region <SAMPLE_REGION>\n          Specify a region for sampling reads from when estimating the threshold\n          probability. If this option is not provided, but --region is provided,\n          the genomic interval passed to --region will be used. Format should be\n          <chrom_name>:<start>-<end> or <chrom_name>\n\nModified Base Options:\n      --motif <MOTIF> <MOTIF>  Output pileup counts for only sequence motifs\n                               provided. The first argument should be the\n                               sequence motif and the second argument is the\n                               0-based offset to the base to pileup base\n                               modification counts for. For example: --motif\n                               CGCG 0 indicates to pileup counts for the first C\n                               on the top strand and the last C (complement to\n                               G) on the bottom strand. The --cpg argument is\n                               short hand for --motif CG 0\n      --cpg                    Only output counts at CpG motifs. Requires a\n                               reference sequence to be provided as well as FAI\n                               index\n  -r, --ref <REFERENCE_FASTA>  Reference sequence in FASTA format. Required for\n                               motif (e.g. CpG) filtering, requires FAI fasta\n                               index to be pre-generated\n      --combine-strands        When performing motif analysis (such as CpG), sum\n                               the counts from the positive and negative strands\n                               into the counts for the positive strand position\n\nOutput Options:\n      --header\n          Output a header with the bedMethyl\n      --prefix <PREFIX>\n          Prefix to prepend on bedgraph output file names. Without this option\n          the files will be <mod_code>_<strand>.bedgraph\n      --partition-tag <PARTITION_TAG>\n          Partition output into multiple bedMethyl files based on tag-value\n          pairs. The output will be multiple bedMethyl files with the format\n          `<prefix>_<tag_value_1>_<tag_value_2>_<tag_value_n>.bed` prefix is\n          optional and set with the `--prefix` flag",
      "metadata": {}
    },
    {
      "id": "2dd5509a-9785-4a38-970c-4897a54ff411",
      "cell_type": "markdown",
      "source": "## To achieve the number of methylations per read we did:\n\nmodkit extract calls \\\n./data/IE20_multiplex/IE20_barcode03/IE20_barcode03_lowfiltered_alignedcut.bam \\\n./output/IE20_multiplex/IE20_barcode03/IE20_barcode03_CpG_island_subtelomeres_6000.tsv \\\n--reference ./data/reference_genomes/cut_hg002v1.1.fasta \\\n--include-bed ./data/reference_regions/CpG_island_subtelomeres_6000.bed \\\n--filter-threshold C:0.80 \\\n--cpg \n\n## Now we use a python script to sum all m calls on those regions and generate another file, with the read_id, number of m, and end site. At some point we should also calculate the non-methylated cpg to the a %methylation on the region. I think we can do this with another python script.\n\nbgzip pipelog.bed\n\nmodkit stats \\\n/storage/scratch01/groups/bu/teloseq/modkit/data/reference_regions/IE20_barcode03_chrreference.bed.gz \\\n--regions /storage/scratch01/groups/bu/teloseq/modkit/data/reference_regions/CpG_island_subtelomeres_3000.bed \\\n--out-table IE20_barcode03_CpGislands_bamreference_stats.tsv\n> bedmethyl does not have contig chr14_MATERNAL_P, skipping.\n> bedmethyl does not have contig chr15_MATERNAL_P, skipping.\n> bedmethyl does not have contig chr15_MATERNAL_Q, skipping.\n> bedmethyl does not have contig chr15_PATERNAL_P, skipping.\n> bedmethyl does not have contig chr16_PATERNAL_Q, skipping.\n> bedmethyl does not have contig chr18_PATERNAL_Q, skipping.\n> bedmethyl does not have contig chr19_PATERNAL_Q, skipping.\n> bedmethyl does not have contig chr20_MATERNAL_P, skipping.\n> bedmethyl does not have contig chr20_PATERNAL_P, skipping.\n> bedmethyl does not have contig chr21_PATERNAL_Q, skipping.\n> bedmethyl does not have contig chr6_MATERNAL_P, skipping.\n> bedmethyl does not have contig chr7_PATERNAL_P, skipping.\n> bedmethyl does not have contig chr8_PATERNAL_Q, skipping.\n> bedmethyl does not have contig chrY_PATERNAL_P, skipping.\n\n## The chromosomes that have been skipped are because no reads are mapped there. YYYYY genialll, ya tengo el número de count_valid_m per read en esas regiones para cada chromosome :)\n#chrom                  start   end     name    strand  count_a count_valid_a   percent_a       count_h count_valid_h   percent_h       count_m count_valid_m   percent_m\nchr10_MATERNAL_P        21221   24221   .       .       24      3063            0.7835455       26      2816            0.92329544      2309    2816            81.995735\nchr10_MATERNAL_Q        27175   30175   .       .       173     9862            1.7542081       24      7449            0.3221909       2853    7449            38.30044\nchr10_PATERNAL_P        20352   23352   .       .       78      14957           0.521495        242     23979           1.0092164       21339   23979           88.990364\n\n\n## To sum all the m that are per read\ncd /storage/scratch01/groups/bu/teloseq/modkit/output/IE20_multiplex/IE20_barcode03/modkit_extractcalls\npython count_m_perread.py IE20_barcode03_CpG_island_subtelomeres_6000.tsv IE20_barcode03_CpG_island_subtelomeres_6000_perread.tsv\n",
      "metadata": {}
    },
    {
      "id": "30faca17-b846-48d3-bc38-e0b79537251a",
      "cell_type": "markdown",
      "source": "## To count how many CG motif are on the reference so thus how many are we expecting to calculate the percentage\n\npython count_CG.py /storage/scratch01/groups/bu/teloseq/modkit/data/reference_genomes/cut_hg002v1.1.fasta /storage/scratch01/groups/bu/teloseq/modkit/data/reference_regions/CpG_island_subtelomeres_3000_hIPSC.bed CpG_island_subtelomeres_3000_counts.tsv\n\n# Esto nos da el fichero .tsv con el número de CG motif que hay en el genoma de reference cut_hgv002 para la region de window the 3000 base pair\n",
      "metadata": {}
    }
  ]
}